File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/Result.java
package com.roxiemobile.networkingapi.network.rest;

// Best way to return status flag and message from a method in Java
// @link http://stackoverflow.com/a/356293

import com.roxiemobile.networkingapi.network.rest.response.ResponseEntity;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

class Result<S extends ResponseEntity<BodyType>, F, BodyType> {

// MARK: - Construction

    Result(@NotNull S success) {
        // Init instance variables
        mState = State.SUCCESS;
        mSuccess = success;
        mFailure = null;
    }

    Result(@NotNull F failure) {
        // Init instance variables
        mState = State.FAILURE;
        mSuccess = null;
        mFailure = failure;
    }

// MARK: - Methods

    public @Nullable S value() {
        return (mState == State.SUCCESS) ? mSuccess : null;
    }

    public @Nullable F error() {
        return (mState == State.FAILURE) ? mFailure : null;
    }

    public boolean isSuccess() {
        return (mState == State.SUCCESS);
    }

// MARK: - Inner Types

    private enum State {
        SUCCESS, FAILURE
    }

// MARK: - Variables

    private final @NotNull State mState;

    private final @Nullable S mSuccess;
    private final @Nullable F mFailure;
}

File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/Result.kt
package com.roxiemobile.networkingapi.network.rest

import com.roxiemobile.networkingapi.network.rest.response.ResponseEntity

// Best way to return status flag and message from a method in Java
// @link http://stackoverflow.com/a/356293

@Deprecated("Must be removed", ReplaceWith("kotlin.Result"))
open class Result<S: ResponseEntity<BodyType>, F, BodyType> {

// MARK: - Construction

    internal constructor(success: S) {
        _state = State.SUCCESS
        _success = success
        _failure = null
    }

    internal constructor(failure: F) {
        _state = State.FAILURE
        _success = null
        _failure = failure
    }

// MARK: - Properties

    val isSuccess: Boolean
        get() = (_state == State.SUCCESS)

// MARK: - Methods

    fun value(): S? {
        return if (_state == State.SUCCESS) _success else null
    }

    fun error(): F? {
        return if (_state == State.FAILURE) _failure else null
    }

// MARK: - Inner Types

    private enum class State {
        SUCCESS, FAILURE
    }

// MARK: - Variables

    private val _state: State

    private val _success: S?
    private val _failure: F?
}
